### PR TITLE
Enforcing dependency whitelists in Maven

### PR DESCRIPTION
**Pull request by JonathanGiles**
Can be run with `mvn -f pom.client.xml enforcer:enforce`

### CODE REVIEW

File: pom.client.xml [hunk 1]
```diff
@@ -408,6 +408,87 @@
...
+                    <exclude>*:*:*:*:runtime</exclude>
+                    <exclude>*:*:*:*:compile</exclude>
+                    <exclude>*:*:*:*:provided</exclude>
+                  </excludes>
+                  <includes>
+                    <include>com.azure:*</include>
+                    <include>io.projectreactor</include>
+                    <include>io.projectreactor.netty</include>
+                    <include>io.netty</include>
+                    <include>com.squareup.okhttp3</include>
```
Comment (alzimmermsft) [inline pom.client.xml:444]: Do we want to limit these inclusion to the Netty and OkHttp clients?

File: pom.client.xml [hunk 2]
```diff
@@ -408,6 +408,87 @@
...
+                    <include>org.slf4j</include>
```
Comment (alzimmermsft) [inline pom.client.xml:445]: Could we specifically scope this down to `org.slf4j:slf4j-api`, I don't want use seeing `slf4j-simple` getting included again.

File: pom.client.xml [hunk 3]
```diff
@@ -408,6 +408,87 @@
...
+                    <include>com.fasterxml.jackson.*</include>
+                    <include>io.opentelemetry</include>
```
Comment (alzimmermsft) [inline pom.client.xml:447]: Should this be scoped to the tracing library only?

File: pom.client.xml [hunk 4]
```diff
@@ -408,6 +408,87 @@
...
+                    <include>net.java.dev.jna</include>
```
Comment (alzimmermsft) [inline pom.client.xml:448]: Mind adding a comment for what this is and where it is being used.

File: pom.client.xml [hunk 5]
```diff
@@ -408,6 +408,87 @@
...
+
+                    <!-- FIXME temporary support for cosmos dependencies -->
+                    <include>io.micrometer:micrometer-core</include>
+                    <include>io.dropwizard.metrics:metrics-core</include>
+                    <include>com.fasterxml.uuid:java-uuid-generator</include>
+
+                    <!-- azure-identity requires the following: -->
+                    <include>com.nimbusds:oauth2-oidc-sdk</include>
+                    <include>com.microsoft.azure:msal4j</include>
+                    <include>org.nanohttpd:nanohttpd</include>
```
Comment (alzimmermsft) [inline pom.client.xml:462]: Could these be moved into their library specific POMs?

File: pom.client.xml [hunk 6]
```diff
@@ -408,6 +408,87 @@
...
+                </requireReleaseDeps>
+
+                <!--
+                Rule: https://maven.apache.org/enforcer/enforcer-rules/requireSameVersions.html
+                This rule is used to enforce that all modules depend on the same version of particular dependencies or
+                plugins.
+                -->
+                <requireSameVersions>
+                  <dependencies>
+                    <dependency>com.azure.*</dependency>
```
Comment (alzimmermsft) [inline pom.client.xml:483]: I don't know if this is going to work all the time as we allow for unreleased dependencies when a module needs it.

File: pom.client.xml [hunk 7]
```diff
@@ -408,6 +408,87 @@
...
+                <requireReleaseDeps>
+                  <message>No Snapshots Allowed!</message>
+                </requireReleaseDeps>
+
+                <!--
+                Rule: https://maven.apache.org/enforcer/enforcer-rules/requireSameVersions.html
+                This rule is used to enforce that all modules depend on the same version of particular dependencies or
+                plugins.
+                -->
+                <requireSameVersions>
```
Comment (alzimmermsft) [inline pom.client.xml:481]: Could we also add the following checks?

https://maven.apache.org/enforcer/enforcer-rules/requireReleaseVersion.html
https://maven.apache.org/enforcer/enforcer-rules/requirePluginVersions.html
https://maven.apache.org/enforcer/enforcer-rules/requirePrerequisite.html
https://maven.apache.org/enforcer/enforcer-rules/requireMavenVersion.html

File: pom.client.xml [hunk 8]
```diff
@@ -408,6 +408,87 @@
...
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-enforcer-plugin</artifactId>
+        <version>3.0.0-M3</version> <!-- {x-version-update;org.apache.maven.plugins:maven-enforcer-plugin;external_dependency} -->
+        <executions>
+          <execution>
+            <id>default-cli</id>
+            <goals>
+              <goal>enforce</goal>
+            </goals>
+            <configuration>
```
Comment (alzimmermsft) [inline pom.client.xml:422]: Would moving this configuration outside of the execution section, into the root of the plugin allow for easier overriding in the inheriting POMs?

File: pom.client.xml (hunk 6 above)
Comment (JimSuplizio) [inline pom.client.xml:483]: @JonathanGiles This is the reason that the analyze step is failing. The configuration for the plugin overloads both dependency and plugins for the requireSameVersion configuration. I've fixed the pom file version scanner as part of [this PR](https://github.com/Azure/azure-sdk-for-java/pull/9787) which has been reviewed and merged into master. Once you update your branch and resubmit the analyze step should pass.

File: pom.client.xml (hunk 2 above)
Comment (JonathanGiles) [reply pom.client.xml:445]: Good point

File: pom.client.xml (hunk 1 above)
Comment (JonathanGiles) [reply pom.client.xml:444]: Done

File: pom.client.xml (hunk 3 above)
Comment (JonathanGiles) [reply pom.client.xml:447]: I didn't really want too many special cases, but this is fair - I'll move some of the special cases down into the specific client library to minimise their use outside of that library.

File: pom.client.xml (hunk 4 above)
Comment (JonathanGiles) [reply pom.client.xml]: I've moved it into the azure-identity whitelist

File: pom.client.xml (hunk 5 above)
Comment (JonathanGiles) [reply pom.client.xml]: Everything has been relocated now.

File: pom.client.xml (hunk 6 above)
Comment (JonathanGiles) [reply pom.client.xml:461]: Yes, I don't particularly care about these other rules - they can be removed when they become a nuisance.

**Pull request merged**